File: KE_C/Mandlebrot/Mandlebrot.c
#include <stdio.h>
#include <stdlib.h>

const char *ASCII = " .:;+=xX$&";
const int ASCII_LEN_MINUS_ONE = 9;

const int WIDTH = 500;
const int HEIGHT = 350;

const float RE_START = -2.0;
const float RE_END = 1.0;
const float IM_START = -1.0;
const float IM_END = 1.0;

const int MAX_ITERS = 500;

int main()
{
    int *iters = malloc(sizeof(int) * WIDTH * HEIGHT);
    if (iters == NULL) {
        fprintf(stderr, "Error allocating iterators list.\n");
        return EXIT_FAILURE;
    }

    for (int y = 0; y < HEIGHT; y++)
    {
        for (int x = 0; x < WIDTH; x++)
        {
            float real = (((float)x / (float)WIDTH) * (RE_END - RE_START)) + RE_START;
            float imaginary = (((float)y / (float)HEIGHT) * (IM_END - IM_START)) + IM_START;

            float re = 0.0;
            float im = 0.0;
            int iter = 0;

            while (re * re + im * im < 4.0 && iter < MAX_ITERS)
            {
                re = re * re - im * im + real;
                im = re * im + re * im + imaginary;
                iter++;
            }

            iters[y * WIDTH + x] = iter;
        }
    }

    int max = 0, min = MAX_ITERS;
    for (int i = 0; i < WIDTH * HEIGHT; i++)
    {
        int iter = iters[i];
        if (iter < MAX_ITERS && iter > max)
        {
            max = iter;
        }
        else if (iter < min)
        {
            min = iter;
        }
    }
    float diff = (float)(max - min);

    char *s = malloc(sizeof(char) * (WIDTH + 1) * HEIGHT + 1);
    if (s == NULL) {
        fprintf(stderr, "Error allocating space for the string.\n");
        return EXIT_FAILURE;
    }
    int row_pos = 0;
    int string_i = 0;
    for (int i = 0; i < WIDTH * HEIGHT; i++)
    {
        int iter = iters[i];

        if (iter == MAX_ITERS)
        {
            s[string_i] = ASCII[ASCII_LEN_MINUS_ONE];
        }
        else
        {
            int mapped = iter - min;
            float pos = (float)mapped / diff;
            s[string_i] = ASCII[(size_t)(pos * (float)ASCII_LEN_MINUS_ONE)];
        }
        string_i++;

        row_pos++;
        if (row_pos == WIDTH)
        {
            row_pos = 0;
            s[string_i] = '\n';
            string_i++;
        }
    }

    s[string_i] = '\0';

    printf("%s", s);

    free(iters);
    free(s);

    return EXIT_SUCCESS;
}

File: KE_Zig/src/mandlebrot.zig
const std = @import("std");

const ASCII = " .:;+=xX$&";
const ASCII_LEN_MINUX_ONE: usize = 9;

const WIDTH: usize = 50;
const HEIGHT: usize = 35;

const WIDTH_X_HEIGHT: usize = WIDTH * HEIGHT;

const RE_START: f32 = -2.0;
const RE_END: f32 = 1.0;
const IM_START: f32 = -1.0;
const IM_END: f32 = 1.0;

const MAX_ITERS: u8 = 50;

pub fn main() !void {
    // var gpa = std.heap.GeneralPurposeAllocator(.{}){};
    // defer _ = gpa.deinit();
    // const allocator = gpa.allocator();

    var iters = [_]u8{0} ** (WIDTH_X_HEIGHT);

    // var x: f32 = 0.0;
    // var y: f32 = 0.0;

    // while (y < HEIGHT) : (y += 1.0) {
    //     while (x < WIDTH) : (x += 1.0) {

    //     }
    // } TODO: check speed of this vs below

    for ([_]u8{0} ** HEIGHT) |_, y| {
        for ([_]u8{0} ** WIDTH) |_, x| {
            const real = (@intToFloat(f32, x) / @as(f32, WIDTH)) * (RE_END - RE_START) + RE_START;
            const imaginary = (@intToFloat(f32, y) / @as(f32, HEIGHT)) * (IM_END - IM_START) + IM_START;

            var re: f32 = 0.0;
            var im: f32 = 0.0;
            var iter: u8 = 0.0;

            while (re * re + im * im < 4.0 and iter < MAX_ITERS) {
                re = re * re - im * im + real;
                im = re * im + re * im + imaginary;
                iter += 1;
            }

            iters[y * WIDTH + x] = iter;
        }
    }

    var max: u8 = 0;
    var min: u8 = MAX_ITERS;

    for (iters) |iter| {
        if (iter < MAX_ITERS and iter > max) {
            max = iter;
        } else if (iter < min) {
            min = iter;
        }
    }

    const diff: f32 = @intToFloat(f32, max - min);

    var string = [_]u8{0} ** ((WIDTH + 1) * HEIGHT + 1);
    var row_pos: usize = 0;
    var string_i: usize = 0;
    for (iters) |iter| {
        if (iter == MAX_ITERS) {
            string[string_i] = ASCII[ASCII_LEN_MINUX_ONE];
        } else {
            const mapped = iter - min;
            const pos = @intToFloat(f32, mapped) / diff;
            string[string_i] = ASCII[@floatToInt(usize, pos * @as(f32, ASCII_LEN_MINUX_ONE))];
        }

        string_i += 1;

        row_pos += 1;
        if (row_pos == WIDTH) {
            row_pos = 0;
            string[string_i] = '\n';
            string_i += 1;
        }
    }

    std.debug.print("{s}", .{string});
}
